elastic/kibana
# [Search] Use session service on a dashboard

## Release note

Adds Search Sessions support to Kibana Dashboards

## Summary

Follow up on https://github.com/elastic/kibana/pull/76889
Prev version of this pr with some comments: https://github.com/elastic/kibana/pull/81192
Part of https://github.com/elastic/kibana/issues/61738
Next step: restore searchSessionId from the URL: https://github.com/elastic/kibana/pull/81489

### Some context for reviewers

This is part of client side background search. To save or restore a background search, searches will be grouped under a session. We implemented session creation on every search in discover in #76889. This pr creates a search session on every dashboard container state change and propagates `searchSessionId` to embeddable -> search infrastructure. 

### Implementation notes:

1. Explicitly passes sessionId as embeddable input: 
  - [x] Visualize
  - [x] Lens
  - [ ] ~Maps: TODO: probably better done separately~ 
  - [ ] ~Anything else?~
2. For now this session id is only needed to group error toasts. For example, see two errors and a single toast. Before this pr you would see two separate toasts. The underlying code for grouping errors was added in https://github.com/elastic/kibana/pull/76889

![Screenshot 2020-10-21 at 17 12 13](https://user-images.githubusercontent.com/7784120/96739888-93d83200-13c0-11eb-8d0a-4f128f1e07b2.png)
3. Added `sessionId` displayed in inspector. Used it it in a functional test. I also think it would be useful for debug and support purposes. 
This is how it looks: 

![Screenshot 2020-10-21 at 17 07 54](https://user-images.githubusercontent.com/7784120/96739842-84f17f80-13c0-11eb-8d58-bc817fa21da0.png)

### Functional test:

Added a test that checks:

1. In case multiple panels failed, there is only single toast 
2. SessionId in inspector is the same for different panels
3. After state change and refetch sessionId changes

### How to test 

* You go to the dashboard and check that `searchSessionId` changes on different state changes. You can use inspector. 


### Open questions / edge cases to iterate over later
* Should refresh interval create a new session Id? - now it doesn't on a dashboard
* Internal embeddable state changes, like own filters, wouldn't create a new sessionId. This is fine.

### Next steps 

* Integrate`searchSessionId` with dashboardURL https://github.com/elastic/kibana/pull/81633 


### Checklist

Delete any items that are not applicable to this PR.
